README.md
# VoyagerNewBooks


CHANGES
New Books List

Copyright 2000-2017, The University of Texas at Arlington

Written by Michael Doran, doran@uta.edu
UTA Libraries, Box 19497, Arlington, TX 76019, USA

See also:  http://rocky.uta.edu/doran/newbooks/

"Voyager" and "WebVoyage" are trademarks of Ex Libris

|------------------------------------------------------------|
| Version 7.6 CHANGES:                                       |
|------------------------------------------------------------|
Improve error message when newbooks.pl can't open the
newBooks.txt file.

|------------------------------------------------------------|
| Version 7.5 CHANGES: Default to en_US skin                 |
|------------------------------------------------------------|

When there is no skin in newBooks.ini %language_modules
corresponding to the current skin, default to the en_US value.

If there en_US is also missing, print a fatal error.


|------------------------------------------------------------|
| Version 7.4 CHANGES: SSL WebVoyage fix                     |
|------------------------------------------------------------|

Method for detecting http vs https was broken.  Use more robust
methods.

Fix error reporting when unable to connect to WebVoyage so that
the output is valid CGI output.

Formatting cleanups to improve legibility.

|------------------------------------------------------------|
| Version 7.3 CHANGES: Tweaks and fixes                      |
|------------------------------------------------------------|

The newBooks.cgi able to play nice in the Ex Libris cloud 
environment, courtesy of code hacking by James Mitchell.

Download package now a zip file rather than a gzipped tar file.

Removed "dope.sh" troubleshooting script from package. 

|------------------------------------------------------------|
| Version 7.0 CHANGES: enhancements                          |
|------------------------------------------------------------|

Voyager 7.2 compatibility
-------------------------
Minor tweaks to work with Tomcat 7.2 changes.

Voyager 7.0 compatibility
-------------------------
Integrates functionally into the Voyager 7.0 server-side 
WebVoyage tomcat architecture.  Implements the look and feel 
and functionality of the new WebVoyage user interface. 

Book cover thumbnails
---------------------
Can enable Google or Syndetic Solutions book cover thumbnail
images.  Thumbnail image is linked to the respective service.
This enhancement is still in 'beta' development.

Google-like keyword search
--------------------------
Words are automatically boolean ANDed.  Words in quotes are
searched as a phrase.

Unicode-aware keyword searching 
-----------------------------
Search terms containing UTF-8 encoded diacritics are handled
correctly.

Keyword search truncation
-------------------------
Ending truncation with an * or a ?.

Search term highlighting
------------------------
Search terms are highlighted in the results pages.


|------------------------------------------------------------|
| Version 6.1.2 CHANGES: enhancement                         |
|------------------------------------------------------------|

Utility revision
----------------
The utility of discovering the Oracle-Perl environment has
been revised and renamed "dope.sh".


|------------------------------------------------------------|
| Version 6.1.1 CHANGES: fix and enhancement                 |
|------------------------------------------------------------|

Path to Perl
------------
Changed the path to Perl on the top line of the scripts to
point to a version of Perl that is being installed by 
Endeavor during the Voyager with Unicode upgrade.  This Perl
version has new DBI and DBD::Oracle modules that are 
compatible with the upgraded Voyager Oracle version. 

Utility for determining Perl versions and DBI/DBD modules
---------------------------------------------------------
Included a utility ("versions.sh") that can be run from the 
command line.  The script attempts to determine how many 
copies of Perl are installed on the system, what versions 
they are, and which ones have the DBI & DBD::Oracle modules.
 

|------------------------------------------------------------|
| Version 6.1 CHANGES: enhancement                           |
|------------------------------------------------------------|

Canned link via bib_id
----------------------
The link-back to the WebVoyage record from the New Books List
result pages is now accomplished via the bib ID redirect that
was enabled as part of the Voyager with Unicode release (and
that was actually implemented at some point for 2001.2). This
enhancement solves the problems inherent in linking back by 
title for the items that lacked an ISBN number. 


|------------------------------------------------------------|
| Version 6.0 CHANGES: enhancement                           |
|------------------------------------------------------------|

Unicode compatability (enhancement)
---------------------
This version of the New Books List has been designed for

dope.sh
#!/usr/bin/sh

#  Shell script: dope.sh - Discover Oracle-Perl Environment
#
#  version 0.9 beta
#
#  2005, Michael Doran, doran@uta.edu
#  University of Texas at Arlington Libraries
#
#  dope - information esp. from a reliable source
#  (Webster's Ninth New Collegiate Dictionary)
#  
#  Quick hack... no warranties, use at your own
#  risk, yada, yada.

dope_version="0.9 beta"

echo "DOPE $dope_version - Discover Oracle-Perl Environment\n"

#
#  Search for Solaris Perl packages
#
if [ -x /usr/bin/pkginfo ]
then
    echo `uname -a` "\n"
    echo "Searching for Solaris Perl packages..."
    pkgs="`pkginfo | grep Perl \
       | awk '{ print \"  \" $2, $3, $4, $5, $6, $7 }'`"
    if [ "$pkgs" ]
    then
        echo "$pkgs\n"
    else
        echo "  None found\n"
    fi
else
    echo "The `basename $0` utility was developed for Solaris..."
    sleep 2
    echo "  ...and you are apparently running it on another system."
    sleep 2
    echo "  Which is probably OK, but your results may vary.\n"
    sleep 5
fi

#
#  Search for 'perl' executables
#
echo "Searching for 'perl' executables..."
for perl in `find / -type f -name "perl" 2>/dev/null`
do
    if [ -x $perl ]
    then
        perl_found="yes"
        echo " " $perl
        echo "    Inode#:" `ls -i $perl | \
            awk '{print $1}'` "\tVersion:" `$perl -v | \
            grep 'This is perl' | \
            sed -e "s/.*[v ]\([\.0-9]\{2,\}\) .*/\1/"`
        modules=`$perl -e 'use ExtUtils::Installed; my $instmod = ExtUtils::Installed->new(); foreach my $module ($instmod->modules()) { my $version = $instmod->version($module) || "???"; print "      $module $version\n"; }'`
        p_modules="`echo \"$modules\" | grep -v Perl`"
        if [ "$p_modules" ]
        then
            echo "    ...has these 'after market' modules:"
            echo "$p_modules" | grep -v "Perl"
        fi
    fi
done
if [ $perl_found ]
then
    echo "  Note: Identical inode numbers indicate hard-linked files." 
else
    echo "  None found\n"
fi
    echo "  Note: There may be Perl executables under a different name.\n" 

#
#  Search for 'perl' symbolic links
#
echo "Searching for Perl symbolic links..."
for perl in `find / -type l -name "perl" 2>/dev/null`
do
if [ -x $perl ]
then
    link="yes"
    echo " " `ls $perl` `ls -l $perl | \
        sed -e "s/.*\(->.*\)/\1/"` "("`$perl -v | \
        grep 'This is perl' | \
        sed -e "s/.*[v ]\([\.0-9]\{2,\}\) .*/\1/"`")"
fi
done
if [ $link ]
then
    echo ""
else
    echo "  None found\n"
fi

#
# Search for Perl DBI module
#
echo "Searching for locations of Perl DBI module(s)..."
msg="None found\n"
for dbi in `find / -type f -name "DBI.pm" 2>/dev/null | grep -v Bundle`
do
    echo " " $dbi
    msg=""
done
echo " " $msg

#
# Search for Perl DBI module
#
echo "Searching for locations of Perl DBD::Oracle module(s)..."
msg="None found\n"
for dbi in `find / -type f -name "Oracle.pm" 2>/dev/null | grep -v Bundle`
do
    echo " " $dbi
    msg=""
done
echo " " $msg

#
# Search for Oracle versions
#
echo "Searching for Oracle versions on this system..."
for i in `find / -name product 2>/dev/null | grep oracle | grep app`
do
    if [ -d $i ]
    then
        oracle_found="yes"
        echo "  $i has these versions..."
        oracle_versions=`ls $i`
        if [ "$oracle_versions" ]
        then
            echo "    " $oracle_versions
        else
            echo "    Hmmmm.  Seems to be empty."
        fi
    fi
done
if [ ! "$oracle_found" ]
then
    echo "  None found\n"
fi


gbsv-jquery.js
/*##########################################################################
# Copyright 2008 Rector and Visitors of the University of Virginia
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################################################################*/

/*

A jQuery plugin for the Google Book Search Viewability API - http://code.google.com/apis/books/getting-started.html

Version: 		0.5.0
Date Modified: 		3-18-08
Author: 		Matt Mitchell - mwm4n

Recent Changes:
	Completely removed overly-complicated "span" tag scheme, in favor of working with the real HTML source.

If the bibkey (a.title or img.alt) is invalid, the element is left alone.
	For links this should always be #preview or #info
	For images, this leaves the option of having a default "src"

Example HTML
	<dt>Thumbnail</dt>
	<dd>
		<img src="default_cover.gif" class="gbsv" alt="ISBN:0743226720"/>
	</dd>
	
	<dt>Preview Link</dt>
	<dd>
		<a href="#preview" class="gbsv" title="ISBN:0743226720">Preview</a>
	</dd>
	
	<dt>Info Link</dt>
	<dd>
		<a href="#info" class="gbsv" title="ISBN:0743226720">Info</a>
	</dd>
	
	<dt>Preview Link + Thumbnail - nested img tag doesn't need bib-key</dt>
	<dd>
		<a href="#preview" class="gbsv" title="ISBN:0743226720"><img src="default_cover.gif" class="gbsv"/></a>
	</dd>
	
	<dt>Info Link + Thumbnail - nested img tag doesn't need bib-key</dt>
	<dd>
		<a href="#info" class="gbsv" title="ISBN:0743226720"><img src="default_cover.gif" class="gbsv"/></a>
	</dd>

Example Javascript
	$(function(){ $.GBSV.init() })

Notes
	Another javascript "Google Book Service Viewability" solution: libx.org/gbs

*/

jQuery.GBSV = {
	
	/* The HTL "class" prefix */
	class_name : 'gbsv',
	
	/* Base url to Google Books */
	url : 'http://books.google.com/books',
	
	/* Base query string for the initial GBS call */
	base_query : 'jscmd=viewapi&callback=jQuery.GBSV.processBookInfoResponse&bibkeys=',
	
	/*
		initializes properties, loops through each "a" and "img" tag that has a "gbsv" class
		collects the bibkey from a.title and img.alt attributes
			- the a tags must have a named anchor of either "preview" or "info"
		builds query for GBSV service request
		appends a script tag with the generated url
		populates the img.src attribute and the a.href attribute
	*/
	init : function(){
		var bibkeys = [];
		var found_elements = this.found_elements={}
		
		jQuery('a.' + this.class_name).each(function(){
			var t=jQuery(this)
			bibkeys.push(t.attr('title'));
			found_elements[t.attr('title')] = found_elements[t.attr('title')] ? found_elements[t.attr('title')] : []
			found_elements[t.attr('title')].push(t);
		});
		
		jQuery('img.' + this.class_name).each(function(){
			var t=jQuery(this);
			if(t.parent()[0].tagName=='A'){return false;}
			bibkeys.push(t.attr('alt'));
			found_elements[t.attr('alt')] = found_elements[t.attr('alt')] ? found_elements[t.attr('alt')] : []
			found_elements[t.attr('alt')].push(t);
		});
		
		this.loadFromBibKeys(bibkeys);
	},
	
	/*
		Creates the request url from the bibkeys arg, then appends a script tag using the url as the src
	*/
	loadFromBibKeys : function(bibkeys){
		jQuery('body').append('<script src=' + this.createRequestURL(bibkeys) + '/>');
	},
	
	/*
		Creates the request url using the bibkeys arg
	*/
	createRequestURL : function(bibkeys){
		return this.url + '?' + this.base_query + this.arrayUnique(bibkeys).join(',')
	},
	
	/*
		Removes duplicate values from an array
	*/
	arrayUnique : function(src_array){
		var o = new Object();
		var i, e;
		for (i = 0; e = src_array[i]; i++) {o[e] = 1};
		var a = new Array();
		for (e in o) {a.push (e)};
		return a;
	},
	
	/*
		Process the data returned from the GBSV service
		Calls handleItem for each item
	*/
	processBookInfoResponse : function(data){
		for(bibkey in this.found_elements){
			for(i in this.found_elements[bibkey]){
				this.processElement(this.found_elements[bibkey][i], data[bibkey]);
			}
		}
	},
	
	processElement : function(el, data){
		if(el[0] && el[0].tagName=='A'){
			this.processAElement(el, data);
		}
		if(el[0] && el[0].tagName=='IMG'){
			this.processImgElement(el, data);
		}
	},
	
	processAElement : function(el, data){
		var k=el.attr('href').match(/[^#]+/);
		el.attr('href', data[k + '_url']);
		// Find an img.gbsv elements that might be inside of this a tag, apply the thumbnail src
		jQuery('img.' + this.class_name, el).attr('src', data['thumbnail_url']);
	},
	
	processImgElement : function(el, data){
		if(data) el.attr('src', data['thumbnail_url']);
	}
	
}

newbooks.pl
#!/m1/shared/bin/perl

########################################################################
#
#  newBooks.pl : a New Books List program
#
#  Version: 7.3 for Unix
#
#  Created by Michael Doran, doran@uta.edu
#
#  University of Texas at Arlington Libraries
#  Box 19497, Arlington, TX 76019, USA
#
#  This Perl program has two distinct parts:
#   1) The first connects to the Voyager database and
#      extracts data on "new" items via an SQL query.
#   2) The second part of the script transfers the
#      flat-file database to where it can be accessed
#      by the New Books CGI program (newBooks.cgi).
#
#  More information at: http://rocky.uta.edu/doran/newbooks/
#
########################################################################
#
#  Copyright 2000-2017, The University of Texas at Arlington ("UTA").
#  All rights reserved.  See included LICENSE for particulars.
#
########################################################################
#
#  "Voyager" and "WebVoyage" are trademarks of Ex Libris
#
########################################################################

use strict;

#  This program requires the DBI and DBD::Oracle modules
#  which are not a part of the default Perl distribution.
#  Since Voyager 2000.1, Ex Libris has installed them.

use DBI;

############################################################
#  Part 1 configuration: extracting the data
############################################################
#
#  Running as a crontab entry requires that we set some
#  environment variables that would already be set if we
#  were running this from the command line while logged in.
#  To see if your values for these are the same, run the
#  "env" command while logged in to your database server
#  as the 'voyager' user.

$ENV{ORACLE_SID} = "VGER" unless exists($ENV{ORACLE_SID}) && defined($ENV{ORACLE_SID});
$ENV{ORACLE_HOME} = "/oracle/app/oracle/product/12.1.0.2/db_1" unless exists($ENV{ORACLE_HOME}) && defined($ENV{ORACLE_HOME});

#  Database name
#  Substitute your database in place of xxxdb.

my $db_name = "xxxdb";

#  Name of the output flat-file of new book records
#  You probably shouldn't change this variable.

my $out_file = "newBooks.txt";

#  Voyager directory prefix
#  You probably won't need to change this variable.

my $dir_prefix = "/m1";

#  Report directory where you will temporarily store the
#  output file prior to transfering it to WebVoyage.
#  You probably won't need to change this variable.

my $report_dir = "$dir_prefix/voyager/$db_name/rpt";

#  Voyager (Oracle) read-only username and password
#  Note: The read-only username and password may be
#        different with the Unicode upgrade.
#        e.g. ro_xxxdb/ro_xxxdb
#  You WILL need to change these variables.

my $username = "ro_xxxdb";
my $password = "ro_xxxdb";


######################################
#  The SQL
#
#  The query that extracts "new" items out of the Voyager database
#  is divided into two parts that are run sequentially:
#   SQL pass 1 - physical items
#   SQL pass 2 - electronic-only items with a mfhd 856
#
#  Details of the logic involved in how newBooks.pl defines "what
#  is a new item" (with respect to the Voyager database tables),
#  can best be determined  by examining the actual SQL code below.


######################################
#  Intervals of weeks or months 
#
#  You have the option of retrieving either 4 weeks
#  or 4 months worth of items.  This option also
#  appears in the newBooks.ini config file.
#  Obviously, THE SELECTION YOU MAKE HERE MUST MATCH
#  the selection in newBooks.ini in order for your
#  results to make sense. 
#
#  This option was added so that smaller libraries
#  that don't add many items per week can have a
#  decent-sized new books database.

#  Choices are weeks or months.

my $intervals = "weeks";


######################################
#  Lag time 
#
#  The default lag time is one week.  This gives
#  your tech services dept that long from the time
#  the item is received (SQL option 1) or the
#  bib record is created (SQL option 2) before
#  that item is pulled for the new books list.  If
#  you notice that too many items are "In Process"
#  you can change the lag time to two weeks. 

#  Choices are 1 or 2. 

my $lag_time = "1";


############################################################
#  Part 2 configuration: transferring the output file
############################################################
#
#  There are three options for transferring the output file
#     1) Copy it to a directory - This is the obvious choice
#        if you run newBooks.pl on the same server that you
#        run the newBooks.cgi program.
#     2) Remote copy it to another server - This is one 
#        method for transferring to a remote server.
#        It requires that you create a ".rhosts" file in the
#        home directory of the voyager user on the remote
#        machine.
#     3) FTP it to another server - this requires the Perl
#        Net::FTP module (which should already be installed).
#        This option requires that you include the username 
#        and password of the voyager user on the remote machine.

#  First select which option (choices are 1, 2, or 3).

my $transfer_option = "1";

#  Then edit the variables for your option choice:

######################################
#  Transfer option 1: copy
#
#  This method should be supported on
#  all Unix platforms.

#  This directory is the final destination
#  (on this server) for the output file.

my $destination_dir = "$dir_prefix/voyager/$db_name/tomcat/vwebv/context/vwebv/newbooks";

######################################
#  Transfer option 2: remote copy
#
#  Remote copy uses the Perl 'system' command to run the native 
#  Unix rcp command.  IN ORDER FOR IT TO WORK you need to create
#  an .rhosts file in the home directory of the voyager user on 
#  the remote machine (i.e. the WebVoyage server).  The .rhosts 
#  file should have the following ownerships and permissions:
#	-rw-------   1 voyager  endeavor  ...   .rhosts
#
#  The /export/home/voyager/.rhosts file you just created should
#  contain an entry consisting of your database server name (the 
#  host name of the server that this script is running on) and the
#  user name (voyager) that is being allowed to do the remote copy.
#  e.g.:
#	database_server_name	voyager
#
#  Important Note: The "database_server_name" specified in the
#  rhosts file must match, exactly, the FIRST listing after the
#  IP address of that server in the /etc/inet/hosts file (or you
#  will get a "permission denied" message when trying to rcp).
#  e.g. If you use a fully qualified domain name in .rhosts:
#  	VYGRDB.UNIV.EDU  voyager
#  then the entry in the hosts file should be:
#	123.45.67.89	VYGRDB.UNIV.EDU  vygrdb  loghost
#  Conversly, if you use just the host name in .rhosts:
#	VYGRDB  voyager
#  then the entry in the hosts file should be:
#	123.45.67.89	VYGRDB  vygrdb.univ.edu  loghost
#  This is a subtle, but important point.
#
#  The remote copy method should work on all Unix platforms 
#  provided you create the .rhosts file and that you run 
#  the newBooks.pl program as the voyager user (or via the
#  voyager crontab). 
#
#  Also, on the WebVoyage server:
#  If you have turned off access to the in.rshd daemon in the
#  /etc/inet/inetd.conf file (by commenting it out) you will
#  have to make it available.  Remember to send a HUP signal
#  to the inetd process if you edit inetd.conf. 
#
#  If you are running TCP_wrappers, and you restrict in.rshd 
#  in inetd.conf, you may have to add the appropriate entry to 
#  the /etc/hosts.allow file of your WebVoyage server: 
#	in.rshd: 123.45.67.89
#  (use the IP address of your Voyager database server.)

my $remote_dir = "/m1/voyager/$db_name/tomcat/vwebv/context/vwebv/newbooks";

my $webvoyage_server = "your.server.edu";

######################################
#  Transfer option 3: FTP
#
#  Since this requires you to supply the voyager username and 
#  password, I recommend that you set permissions on this script
#  to 700 (i.e. chmod 700 newBooks.pl).  The voyager user should
#  be the owner of the script.

my $ftpuser = "voyager";
my $passwd  = "xxxxxxx";

# If using option 3 (i.e. FTP), you must uncomment the following line.

# use Net::FTP;

  #       *** NOTE *** NOTE *** NOTE ***		   #
  #	The $webvoyage_server and $remote_dir variables	   #
  #	for transfer option 2 are ALSO USED for FTP        #
  #	transfers and MUST have values assigned if you     #
  #	use tranfer option 3. 				   # 


########################################################################
#
#    * * * * * * * *                          * * * * * * * *
#    * * * * * * * *    Stop editing here!    * * * * * * * *
#    * * * * * * * *                          * * * * * * * *
#
########################################################################
#
#  Most Voyager sites should not have to edit code beyond this point.
#
########################################################################


################################################################
#
#  Part 1: extracting the data
#
################################################################

#  This gets the ball rolling for part 1.

&DoQuery;

##########################################################
#  DoQuery
##########################################################
#

sub DoQuery {
    # Connect to Oracle database
    my $dbh = DBI->connect('dbi:Oracle:', $username, $password)
	|| die "Could not connect: $DBI::errstr";

    # Prepare the first SQL statement
    my $sth = $dbh->prepare(&ConstructSQL("1")) 
	|| die $dbh->errstr;

    # Run the SQL query
    $sth->execute
	|| die $dbh->errstr;

    # Output the results to a file
    open (OUTFILE, ">$report_dir/$out_file")
	|| die "Cannot create/open output file $report_dir/$out_file: $!";

    while( my (@entry) = $sth->fetchrow_array() ) {
	if ($entry[0]) { 
            # bib_id
	    print OUTFILE "$entry[0]\t"; 
	} else { 
	    print OUTFILE "\t"; 
	}
	if ($entry[1]) { 
            # isbn 
            my $isbn = $entry[1];
            $isbn =~ s/[^\d]*([\d|\-|X]*)[\s|\[|\(]*.*/$1/;
	    print OUTFILE "$isbn\t"; 
	} else { 
	    print OUTFILE "\t"; 
	}
	if ($entry[2]) {
            # bib_format 
	    print OUTFILE "$entry[2]\t"; 
	} else {
	    print OUTFILE "\t";
	}
	if ($entry[3]) { 
            # author 
	    print OUTFILE "$entry[3]\t"; 
	} else { 
	    print OUTFILE "\t"; 
	}
	if ($entry[4]) { 
            # title 
	    print OUTFILE "$entry[4]\t"; 
	} else { 
	    print OUTFILE "\t"; 
	}
	if ($entry[5]) { 
            # edition 
	    print OUTFILE "$entry[5]\t"; 
	} else { 
	    print OUTFILE "\t"; 
	} if ($entry[6]) { 
            # imprint 
	    print OUTFILE "$entry[6]\t"; 
	} else { 
	    print OUTFILE "\t"; 
	}
	if ($entry[7]) { 
            # permanent location 
	    print OUTFILE "$entry[7]\t"; 
	} else { 
	    print OUTFILE "\t"; 
	}
	if ($entry[8]) { 
            # temporary location 
	    print OUTFILE "$entry[8]\t"; 
	} else { 
	    print OUTFILE "\t"; 
	}
	if ($entry[9]) { 
            # display call number
	    print OUTFILE "$entry[9]\t"; 
	} else { 
            # Note: In Process entries deliberately have
            # a leading space for sorting purposes.
            print OUTFILE " In Process\t";
	}
	if ($entry[10]) { 
            # normalized call number 
            print OUTFILE "$entry[10]\t"; 
	} else { 
            print OUTFILE " In Process\t";
	}
	if ($entry[11]) { 
	    print OUTFILE "$entry[11]\n"; 
	} else { 
	    print OUTFILE "$entry[12]\n"; 
	}
    }

    # Prepare the second SQL statement
    $sth = $dbh->prepare(&ConstructSQL("2")) 
	|| die $dbh->errstr;

    # Run the SQL query
    $sth->execute
	|| die $dbh->errstr;


    # Output the results to a file
    open (OUTFILE, ">>$report_dir/$out_file")
	|| die "Cannot create/open output file: $!";

    while( my (@entry) = $sth->fetchrow_array() ) {
	if ($entry[0]) { 
            # bib_id
	    print OUTFILE "$entry[0]\t"; 
	} else { 
	    print OUTFILE "\t"; 
	}
	if ($entry[1]) { 
            # isbn 
            my $isbn = $entry[1];
            $isbn =~ s/[^\d]*([\d|\-|X]*)[\s|\[|\(]*.*/$1/;
	    print OUTFILE "$isbn\t"; 
	} else { 
	    print OUTFILE "\t"; 
	}
	if ($entry[2]) {
            # bib_format 
	    print OUTFILE "$entry[2]\t"; 
	} else {
	    print OUTFILE "\t";
	}
	if ($entry[3]) { 
            # author 
	    print OUTFILE "$entry[3]\t"; 
	} else { 
	    print OUTFILE "\t"; 
	}
	if ($entry[4]) { 
            # title 
	    print OUTFILE "$entry[4]\t"; 
	} else { 
	    print OUTFILE "\t"; 
	}
	if ($entry[5]) { 
            # edition 
	    print OUTFILE "$entry[5]\t"; 
	} else { 
	    print OUTFILE "\t"; 
	} if ($entry[6]) { 
            # imprint 
	    print OUTFILE "$entry[6]\t"; 
	} else { 
	    print OUTFILE "\t"; 
	}
	if ($entry[7]) { 
            # permanent location 
	    print OUTFILE "$entry[7]\t"; 
	} else { 
	    print OUTFILE "\t"; 
	}
        print OUTFILE "\t";
	if ($entry[8]) { 
            # display call number
	    print OUTFILE "$entry[8]\t"; 
	} else { 
            # Note: In Process entries deliberately have
            # a leading space for sorting purposes.
            print OUTFILE "\t";
	}
	if ($entry[9]) { 
            # normalized call number 
	    print OUTFILE "$entry[9]\t"; 
	} else { 
            print OUTFILE "\t";
	}
	if ($entry[10]) { 
	    print OUTFILE "$entry[10]\n"; 
	} else { 
	    print OUTFILE "\n"; 
	}
    }

    $sth->finish;

    $dbh->disconnect;

    close (OUTFILE);

}



##########################################################
#  SetInterval
##########################################################

sub SetInterval {
    my ($one_or_two) = @_;
    my ($date_condition_one,$date_condition_two);
    if ($intervals eq "weeks") {
	if ($lag_time eq "1") {
	    $date_condition_one = " / 7) - 1)";
	    $date_condition_two = "(sysdate - 35) and (sysdate - 7)";
        } elsif ($lag_time eq "2") {
	    $date_condition_one = " / 7) - 2)";
	    $date_condition_two = "(sysdate - 42) and (sysdate - 14)";
	} else {
	    print "Error: no lag time option selected.\n";
	    exit (22);
	}
    } elsif ($intervals eq "months") {
	if ($lag_time eq "1") {
	    $date_condition_one = "-  7) / 30) ";
	    $date_condition_two = "(sysdate - 127) and (sysdate - 7)";
        } elsif ($lag_time eq "2") {
	    $date_condition_one = "- 14) / 30) ";
	    $date_condition_two = "(sysdate - 134) and (sysdate - 14)";
	} else {
	    print "Error: no lag time option selected.\n";
	    exit (22);
	}
    } else {
	print "Error: no interval option selected.\n";
	exit (22);
    }
    if ($one_or_two eq "one") {
	return($date_condition_one);
    } elsif ($one_or_two eq "two") {
	return($date_condition_two);
    }
}


##########################################################
#  ConstructSQL
##########################################################
#
#  This routine returns an SQL query according to which
#  option was selected in the configuration section.

sub ConstructSQL {
    my ($sql_pass) = @_;
    #  The SQL option "choice" is an artifact of previous
    #  versions.  This SQL query's embedded logic as to
    #  "what constitutes a new item" should be adequate
    #  for all sites.  If it isn't, you may create your
    #  own; however, it must output the same fields in
    #  order for the newBooks.cgi to utilize it.  -mdd
    my $date_condition_one = &SetInterval("one");
    my $date_condition_two = &SetInterval("two");
    if ($sql_pass eq "1") {
	return ("
	select distinct
	  $db_name.bib_text.bib_id,
	  $db_name.bib_text.isbn,
	  $db_name.bib_text.bib_format,
	  $db_name.bib_text.author,
	  $db_name.bib_text.title,
	  $db_name.bib_text.edition,
	  $db_name.bib_text.imprint,
	  $db_name.permanent.location_display_name,
	  $db_name.temporary.location_display_name,
	  $db_name.mfhd_master.display_call_no,
	  $db_name.mfhd_master.normalized_call_no,
	  (ceil (((sysdate - $db_name.mfhd_master.update_date) $date_condition_one),
	  (ceil (((sysdate - $db_name.mfhd_master.create_date) $date_condition_one)
	from
	  $db_name.bib_master,
	  $db_name.bib_text,
	  $db_name.bib_mfhd,
	  $db_name.mfhd_item,
	  $db_name.mfhd_master,
	  $db_name.item,
	  $db_name.location permanent,
	  $db_name.location temporary
	where
	  $db_name.bib_master.bib_id=$db_name.bib_text.bib_id and
	  $db_name.bib_text.bib_id=$db_name.bib_mfhd.bib_id and
	  $db_name.bib_mfhd.mfhd_id=$db_name.mfhd_master.mfhd_id and
	  $db_name.mfhd_master.mfhd_id=$db_name.mfhd_item.mfhd_id and
	  $db_name.mfhd_item.item_id=$db_name.item.item_id and
	  $db_name.item.perm_location=$db_name.permanent.location_id and
	  $db_name.item.temp_location=$db_name.temporary.location_id(+) and
	  $db_name.mfhd_master.suppress_in_opac not in ('Y') and
	  $db_name.bib_master.suppress_in_opac not in ('Y') and
	  $db_name.item.on_reserve not in ('Y') and
	  substr($db_name.bib_text.bib_format,-1,1) in ('a','c','m') and
	  $db_name.item.create_date between $date_condition_two and 
	  ($db_name.mfhd_master.create_date between $date_condition_two or
	   $db_name.mfhd_master.update_date between $date_condition_two)
	");
    } elsif ($sql_pass eq "2") {
	return ("
	select distinct
	  $db_name.bib_text.bib_id,
	  $db_name.bib_text.isbn,
	  $db_name.bib_text.bib_format,
	  $db_name.bib_text.author,
	  $db_name.bib_text.title,
	  $db_name.bib_text.edition,
	  $db_name.bib_text.imprint,
	  $db_name.location.location_display_name,
	  $db_name.mfhd_master.display_call_no,
	  $db_name.mfhd_master.normalized_call_no,
	  (ceil (((sysdate - $db_name.mfhd_master.create_date) $date_condition_one)
	from
	  $db_name.bib_master,
	  $db_name.bib_text,
	  $db_name.bib_mfhd,
	  $db_name.mfhd_master,
	  $db_name.mfhd_item,
	  $db_name.location,
	  $db_name.elink_index
	where
	  $db_name.bib_master.bib_id=$db_name.bib_text.bib_id and
	  $db_name.bib_text.bib_id=$db_name.bib_mfhd.bib_id and
	  $db_name.bib_mfhd.mfhd_id=$db_name.mfhd_master.mfhd_id and
	  $db_name.mfhd_master.location_id=$db_name.location.location_id and
	  $db_name.mfhd_master.mfhd_id=$db_name.elink_index.record_id and
	  $db_name.elink_index.record_type in ('M') and
	  $db_name.mfhd_master.mfhd_id=$db_name.mfhd_item.mfhd_id(+) and
	  $db_name.mfhd_item.item_id is null and
	  $db_name.mfhd_master.suppress_in_opac not in ('Y') and
	  $db_name.bib_master.suppress_in_opac not in ('Y') and
	  $db_name.mfhd_master.create_date between $date_condition_two and
	  $db_name.elink_index.link is not null
	");
    } else {
	print "Error: no SQL option selected.\n";
	exit (22);
    }
}


################################################################
#
#  Part 2: transferring the output file
#
################################################################

if      ($transfer_option eq "1" ) {
    &CopyFile;
} elsif ($transfer_option eq "2" ) {
    &RemoteCopyFile;
} elsif ($transfer_option eq "3" ) {
    &SendFTP;
} else {
    print "Aborted file transmission: Incorrect transfer option chosen.\n";
}


##########################################################
#  CopyFile
##########################################################
#
#  This routine copies the output file to the directory
#  on this server where you keep newBooks.cgi.

sub CopyFile {
    use File::Copy;
#	or die "Aborted file transmission: no File::Copy";
    copy("$report_dir/$out_file","$destination_dir/$out_file")
	or die "Aborted file transmission: Can't copy file.";
}


##########################################################
#  RemoteCopyFile
##########################################################
#
#  This routine does a remote copy of the output file to
#  the directory on the server where newBooks.cgi is.

sub RemoteCopyFile {
    system("/usr/bin/rcp $report_dir/$out_file $webvoyage_server:$remote_dir/$out_file");
}


##########################################################
#  SendFTP
##########################################################
#
#  This routine sends the output file to another serve
#  via FTP.  It REQUIRES that you have the Perl libnet
#  module installed so that you can use Net::FTP.

sub SendFTP {
    # Change to directory of output file
    chdir($report_dir)
	or die "FTP: Cannot cd to $report_dir.";

    # Start FTP: connect, and login
    my $ftp = Net::FTP->new($webvoyage_server)
	or die "FTP: Couldn't connect: $@.";
    $ftp->login($ftpuser, $passwd)
	or die "FTP: Couldn't login.";

    # Change to desired webserver directory
    $ftp->cwd($remote_dir)
	or die "FTP: Couldn't change directory.";

    # Send file
    $ftp->put($out_file)
	or die "FTP: Couldn't send $out_file.";

    # Exit from the ftp session
    $ftp->quit()
	or die "FTP: Couldn't quit.";
}


############################################################
# Exit from script

exit(0);
